File: 02-restaurant-waitlist/frontend/src/services/waitlistService.ts
/**
 * Backend-shaped service contract for the waitlist prototype.
 *
 * These types are intentionally explicit and serializable: a later phase will
 * translate them into an OpenAPI contract and a FastAPI backend. React code
 * must only ever talk to the `WaitlistService` interface.
 */

export type PartySizeClass = "A" | "B" | "C" | "D";

export type WaitlistStatus =
  | "pending"
  | "waiting"
  | "notified"
  | "seated"
  | "completed"
  | "cancelled"
  | "no_show";

export type EntryMode = "automatic" | "staff_review";

export interface VenueMessageSettings {
  joinMessage: string;
  policyAcknowledgementMessage: string;
  confirmationMessage: string;
  tableReadyMessage: string;
  cancellationMessage: string;
  closedWaitlistMessage: string;
  largePartyMessage: string;
  largePartyConfirmationMessage: string;
}

export interface Venue {
  id: string;
  name: string;
  logoPlaceholderLabel: string;
  contactPhone: string;
  staffNotificationEmail: string;
  menuUrl: string;
  maxOnlinePartySize: number;
  gracePeriodMinutes: number;
  entryMode: EntryMode;
  waitlistOpen: boolean;
  defaultWaitEstimateMinutes: Record<PartySizeClass, number>;
  messages: VenueMessageSettings;
  serviceDate: string; // YYYY-MM-DD
}

export interface StaffSession {
  token: string;
  staffName: string;
  username: string;
  venueId: string;
  issuedAt: string;
  mocked: true;
}

export type TableAvailabilityState = "available" | "occupied" | "needs_tidying";

export interface Table {
  id: string;
  name: string;
  minCapacity: number;
  maxCapacity: number;
  active: boolean;
  availabilityState: TableAvailabilityState;
  notes: string | null;
  occupyingTicketCode: string | null;
  occupyingEntryId: string | null;
}

export interface CreateTableRequest {
  name: string;
  minCapacity: number;
  maxCapacity: number;
  notes?: string | undefined;
}

export interface UpdateTableRequest {
  name?: string;
  minCapacity?: number;
  maxCapacity?: number;
  active?: boolean;
  notes?: string | null;
}

export interface SetTableAvailabilityRequest {
  availabilityState: "available" | "needs_tidying";
}

export interface WaitlistEntry {
  id: string;
  venueId: string;
  serviceDate: string;
  ticketCode: string;
  ticketSequence: number;
  guestName: string;
  partySize: number;
  originalPartySize: number;
  originalPartyClass: PartySizeClass;
  currentSeatingClass: PartySizeClass;
  reviewRequired: boolean;
  reviewReason: string | null;
  mobileNumber: string;
  seatingNote: string | null;
  status: WaitlistStatus;
  arrivalTime: string;
  estimatedWaitMinutes: number;
  notifiedAt: string | null;
  returnByAt: string | null;
  seatedAt: string | null;
  completedAt: string | null;
  cancelledAt: string | null;
  tableId: string | null;
  tableName: string | null;
  seatingOverrideReason: string | null;
  accessToken: string;
}

export interface LargePartyEnquiry {
  id: string;
  reference: string;
  venueId: string;
  guestName: string;
  partySize: number;
  mobileNumber: string;
  note: string | null;
  createdAt: string;
  message: string;
}

export type NotificationChannel = "sms" | "email" | "in_app";
export type NotificationRecipientType = "guest" | "staff";
export type NotificationTemplateType =
  | "table_ready"
  | "confirmation"
  | "cancellation"
  | "large_party_staff_alert";
export type NotificationDeliveryStatus = "queued" | "sent" | "failed";

export interface Notification {
  id: string;
  entryId: string | null;
  recipientType: NotificationRecipientType;
  channel: NotificationChannel;
  templateType: NotificationTemplateType;
  renderedMessage: string;
  deliveryStatus: NotificationDeliveryStatus;
  createdAt: string;
  sentAt: string | null;
  errorMessage: string | null;
}

export interface ActivityEvent {
  id: string;
  entryId: string | null;
  type: string;
  description: string;
  createdAt: string;
}

export interface WaitlistFilters {
  status?: WaitlistStatus | "all";
  partySizeClass?: PartySizeClass | "all";
  reviewRequiredOnly?: boolean;
  needsAttentionOnly?: boolean;
  serviceDate?: string;
}

export interface CreateGuestEntryRequest {
  guestName: string;
  partySize: number;
  mobileNumber: string;
  seatingNote?: string | undefined;
  policyAcknowledged: boolean;
}

export type CreateGuestEntryResult =
  | {
      kind: "entry";
      entry: WaitlistEntry;
      accessToken: string;
      message: string;
    }
  | {
      kind: "large_party_enquiry";
      enquiry: LargePartyEnquiry;
      message: string;
    }
  | {
      kind: "closed";
      message: string;
    };

export interface GuestStatus {
  ticketCode: string;
  guestName: string;
  partySize: number;
  status: WaitlistStatus;
  estimatedWaitMinutes: number;
  message: string;
  notifiedAt: string | null;
  returnByAt: string | null;
  needsAttention: boolean;
  cancellationAllowed: boolean;
  venueName: string;
  venueLogoPlaceholderLabel: string;
  serviceDate: string;
}

export interface SeatEntryRequest {
  tableId: string;
  seatingOverrideReason?: string | undefined;
}

export interface UpdateVenueRequest {
  name?: string;
  contactPhone?: string;
  staffNotificationEmail?: string;
  menuUrl?: string;
  maxOnlinePartySize?: number;
  gracePeriodMinutes?: number;
  entryMode?: EntryMode;
  waitlistOpen?: boolean;
  defaultWaitEstimateMinutes?: Partial<Record<PartySizeClass, number>>;
  messages?: Partial<VenueMessageSettings>;
}

export interface NotificationResult {
  entry: WaitlistEntry;
  notification: Notification;
}

export interface LoginRequest {
  username: string;
  password: string;
}

export interface LargePartyEnquiryRequest {
  guestName: string;
  partySize: number;
  mobileNumber: string;
  note?: string | undefined;
}

export interface DashboardSummary {
  activeWaiting: number;
  pendingReview: number;
  notified: number;
  needsAttention: number;
}

export interface DashboardData {
  venue: Venue;
  serviceDate: string;
  summary: DashboardSummary;
  entries: WaitlistEntry[];
  tables: Table[];
}

export interface ServiceError extends Error {
  code: string;
}

export interface WaitlistService {
  login(input: LoginRequest): Promise<StaffSession>;
  getVenue(): Promise<Venue>;
  updateVenue(input: UpdateVenueRequest): Promise<Venue>;
  getDashboard(serviceDate?: string): Promise<DashboardData>;
  listWaitlistEntries(filters?: WaitlistFilters): Promise<WaitlistEntry[]>;
  createGuestEntry(input: CreateGuestEntryRequest): Promise<CreateGuestEntryResult>;
  getGuestEntry(accessToken: string): Promise<GuestStatus>;
  cancelGuestEntry(accessToken: string): Promise<WaitlistEntry>;
  approveEntry(entryId: string): Promise<WaitlistEntry>;
  cancelEntry(entryId: string): Promise<WaitlistEntry>;
  notifyEntry(entryId: string): Promise<NotificationResult>;
  returnToWaiting(entryId: string): Promise<WaitlistEntry>;
  extendReturnBy(entryId: string, additionalMinutes: number): Promise<WaitlistEntry>;
  updateWaitEstimate(entryId: string, estimatedWaitMinutes: number): Promise<WaitlistEntry>;
  seatEntry(entryId: string, input: SeatEntryRequest): Promise<WaitlistEntry>;
  markNoShow(entryId: string): Promise<WaitlistEntry>;
  completeEntry(entryId: string): Promise<WaitlistEntry>;
  listTables(): Promise<Table[]>;
  listCompatibleTables(entryId: string): Promise<Table[]>;
  createTable(input: CreateTableRequest): Promise<Table>;
  updateTable(tableId: string, input: UpdateTableRequest): Promise<Table>;
  deleteTable(tableId: string): Promise<void>;
  setTableAvailability(tableId: string, input: SetTableAvailabilityRequest): Promise<Table>;
  createLargePartyEnquiry(input: LargePartyEnquiryRequest): Promise<LargePartyEnquiry>;
  listNotifications(): Promise<Notification[]>;
  createNotificationEvent(input: {
    entryId: string | null;
    recipientType: NotificationRecipientType;
    channel: NotificationChannel;
    templateType: NotificationTemplateType;
    renderedMessage: string;
  }): Promise<Notification>;
}

export function classifyPartySize(partySize: number): PartySizeClass {
  if (partySize <= 2) return "A";
  if (partySize <= 4) return "B";
  if (partySize <= 6) return "C";
  return "D";
}

export function isNeedsAttention(entry: WaitlistEntry, now: Date = new Date()): boolean {
  return (
    entry.status === "notified" &&
    entry.returnByAt !== null &&
    new Date(entry.returnByAt).getTime() < now.getTime()
  );
}

export const PARTY_CLASS_LABELS: Record<PartySizeClass, string> = {
  A: "A (1-2)",
  B: "B (3-4)",
  C: "C (5-6)",
  D: "D (7-12)",
};

export const STATUS_LABELS: Record<WaitlistStatus, string> = {
  pending: "Pending review",
  waiting: "Waiting",
  notified: "Notified",
  seated: "Seated",
  completed: "Completed",
  cancelled: "Cancelled",
  no_show: "No show",
};

File: 02-restaurant-waitlist/backend/services/waitlist_service.py
from datetime import datetime, timedelta

from backend.errors import ServiceError
from backend.models import Notification, Table, WaitlistEntry
from backend.notifications import NotificationProvider
from backend.repository import InMemoryRepository
from backend.schemas import (
    CreateGuestEntryRequest,
    CreateGuestEntryResult,
    CreateGuestEntryResultClosed,
    CreateGuestEntryResultEntry,
    CreateGuestEntryResultLargePartyEnquiry,
    DashboardData,
    DashboardSummary,
    GuestStatus,
    LargePartyEnquiryRequest,
    SeatEntryRequest,
)
from backend.services import large_party_service
from backend.utils import (
    classify_party_size,
    is_needs_attention,
    is_table_compatible,
    is_valid_mobile_number,
    iso_at_offset,
    now_iso,
    sort_entries,
)

ACTIVE_CANCELLABLE_STATUSES = {"pending", "waiting", "notified"}
CLOSED_STATUSES = {"completed", "cancelled", "no_show"}


def find_entry_or_404(repo: InMemoryRepository, entry_id: str) -> WaitlistEntry:
    entry = repo.find_entry(entry_id)
    if entry is None:
        raise ServiceError(404, "entry_not_found", "Waitlist entry not found.")
    return entry


def find_entry_by_token_or_404(repo: InMemoryRepository, access_token: str) -> WaitlistEntry:
    entry = repo.find_entry_by_access_token(access_token)
    if entry is None:
        raise ServiceError(404, "entry_not_found", "We couldn't find this waitlist entry.")
    return entry


def _release_table(repo: InMemoryRepository, entry: WaitlistEntry) -> None:
    if not entry.tableId:
        return
    table = repo.find_table(entry.tableId)
    if table is not None:
        table.availabilityState = "needs_tidying"
        table.occupyingEntryId = None
        table.occupyingTicketCode = None


def get_dashboard(repo: InMemoryRepository, service_date: str | None) -> DashboardData:
    date = service_date or repo.venue.serviceDate
    entries = sort_entries([e for e in repo.entries if e.serviceDate == date])
    summary = DashboardSummary(
        activeWaiting=sum(1 for e in entries if e.status == "waiting"),
        pendingReview=sum(1 for e in entries if e.status == "pending"),
        notified=sum(1 for e in entries if e.status == "notified"),
        needsAttention=sum(1 for e in entries if is_needs_attention(e)),
    )
    return DashboardData(
        venue=repo.venue, serviceDate=date, summary=summary, entries=entries, tables=repo.tables
    )


def list_entries(
    repo: InMemoryRepository,
    *,
    status: str | None,
    party_size_class: str | None,
    review_required_only: bool | None,
    needs_attention_only: bool | None,
    service_date: str | None,
) -> list[WaitlistEntry]:
    date = service_date or repo.venue.serviceDate
    entries = [e for e in repo.entries if e.serviceDate == date]
    if status and status != "all":
        entries = [e for e in entries if e.status == status]
    if party_size_class and party_size_class != "all":
        entries = [e for e in entries if e.currentSeatingClass == party_size_class]
    if review_required_only:
        entries = [e for e in entries if e.reviewRequired]
    if needs_attention_only:
        entries = [e for e in entries if is_needs_attention(e)]
    return sort_entries(entries)


def create_guest_entry(
    repo: InMemoryRepository,
    payload: CreateGuestEntryRequest,
    notifier: NotificationProvider,
) -> CreateGuestEntryResult:
    venue = repo.venue
    if not venue.waitlistOpen:
        return CreateGuestEntryResultClosed(message=venue.messages.closedWaitlistMessage)

    if not payload.policyAcknowledged:
        raise ServiceError(
            400, "policy_not_acknowledged", "You must acknowledge the waitlist policy before joining."
        )
    if not payload.guestName.strip():
        raise ServiceError(400, "invalid_name", "Please enter a name for the party.")
    if payload.partySize < 1:
        raise ServiceError(400, "invalid_party_size", "Party size must be a whole number of one or more.")
    if not is_valid_mobile_number(payload.mobileNumber):
        raise ServiceError(
            400,
            "invalid_mobile_number",
            "Please enter a valid mobile number, for example +61 400 000 000.",
        )

    if payload.partySize > venue.maxOnlinePartySize:
        enquiry = large_party_service.create_enquiry(
            repo,
            LargePartyEnquiryRequest(
                guestName=payload.guestName.strip(),
                partySize=payload.partySize,
                mobileNumber=payload.mobileNumber.strip(),
                note=payload.seatingNote,
            ),
            notifier,
        )
        return CreateGuestEntryResultLargePartyEnquiry(
            enquiry=enquiry, message=venue.messages.largePartyConfirmationMessage
        )

    date = venue.serviceDate
    seq = repo.next_ticket_sequence(date)
    party_class = classify_party_size(payload.partySize)
    entry = WaitlistEntry(
        id=repo.next_id("entry"),
        venueId=venue.id,
        serviceDate=date,
        ticketCode=f"{party_class}-{seq:03d}",
        ticketSequence=seq,
        guestName=payload.guestName.strip(),
        partySize=payload.partySize,
        originalPartySize=payload.partySize,
        originalPartyClass=party_class,
        currentSeatingClass=party_class,
        reviewRequired=False,
        mobileNumber=payload.mobileNumber.strip(),
        seatingNote=payload.seatingNote.strip() if payload.seatingNote and payload.seatingNote.strip() else None,
        status="pending" if venue.entryMode == "staff_review" else "waiting",
        arrivalTime=now_iso(),
        estimatedWaitMinutes=venue.defaultWaitEstimateMinutes[party_class],
        accessToken=repo.next_access_token(),
    )
    repo.entries.append(entry)

    notification = Notification(
        id=repo.next_id("notif"),
        entryId=entry.id,
        recipientType="guest",
        channel="in_app",
        templateType="confirmation",
        renderedMessage=venue.messages.confirmationMessage,
        deliveryStatus="sent",
        createdAt=now_iso(),
        sentAt=now_iso(),
    )
    repo.notifications.append(notification)
    notifier.send(notification)

    return CreateGuestEntryResultEntry(
        entry=entry, accessToken=entry.accessToken, message=venue.messages.confirmationMessage
    )


def _guest_message(repo: InMemoryRepository, entry: WaitlistEntry) -> str:
    m = repo.venue.messages
    if entry.status == "notified":
        return m.tableReadyMessage
    if entry.status == "cancelled":
        return m.cancellationMessage
    if entry.status == "seated":
        return f"You're seated at {entry.tableName or 'your table'}. Enjoy your meal."
    if entry.status == "completed":
        return "Thanks for dining with us today."
    if entry.status == "no_show":
        return "We weren't able to seat this party. Please talk to the host stand."
    if entry.status == "pending":
        return "Your request is with our host stand for review."
    return m.confirmationMessage


def get_guest_entry(repo: InMemoryRepository, access_token: str) -> GuestStatus:
    entry = find_entry_by_token_or_404(repo, access_token)
    return GuestStatus(
        ticketCode=entry.ticketCode,
        guestName=entry.guestName,
        partySize=entry.partySize,
        status=entry.status,
        estimatedWaitMinutes=entry.estimatedWaitMinutes,
        message=_guest_message(repo, entry),
        notifiedAt=entry.notifiedAt,
        returnByAt=entry.returnByAt,
        needsAttention=is_needs_attention(entry),
        cancellationAllowed=entry.status in ACTIVE_CANCELLABLE_STATUSES,
        venueName=repo.venue.name,
        venueLogoPlaceholderLabel=repo.venue.logoPlaceholderLabel,
        serviceDate=entry.serviceDate,
    )


def cancel_guest_entry(
    repo: InMemoryRepository, access_token: str, notifier: NotificationProvider
) -> WaitlistEntry:
    entry = find_entry_by_token_or_404(repo, access_token)
    if entry.status not in ACTIVE_CANCELLABLE_STATUSES:
        raise ServiceError(
            400, "cancellation_not_allowed", "This entry can no longer be cancelled online."
        )
    entry.status = "cancelled"
    entry.cancelledAt = now_iso()
    notification = Notification(
        id=repo.next_id("notif"),
        entryId=entry.id,
        recipientType="staff",
        channel="in_app",
        templateType="cancellation",
        renderedMessage=f"{entry.ticketCode} ({entry.guestName}) cancelled their waitlist entry.",
        deliveryStatus="sent",
        createdAt=now_iso(),
        sentAt=now_iso(),
    )
    repo.notifications.append(notification)
    notifier.send(notification)
    return entry


def approve_entry(repo: InMemoryRepository, entry_id: str) -> WaitlistEntry:
    entry = find_entry_or_404(repo, entry_id)
    if entry.status != "pending":
        raise ServiceError(400, "invalid_status", "Only pending entries can be approved.")
    entry.status = "waiting"
    return entry


def cancel_entry(repo: InMemoryRepository, entry_id: str) -> WaitlistEntry:
    entry = find_entry_or_404(repo, entry_id)
    if entry.status in CLOSED_STATUSES:
        raise ServiceError(400, "invalid_status", "This entry is already closed.")
    _release_table(repo, entry)
    entry.status = "cancelled"
    entry.cancelledAt = now_iso()
    return entry


def notify_entry(
    repo: InMemoryRepository, entry_id: str, notifier: NotificationProvider
) -> tuple[WaitlistEntry, Notification]:
    entry = find_entry_or_404(repo, entry_id)
    if entry.status not in {"waiting", "notified"}:
        raise ServiceError(400, "invalid_status", "Only waiting guests can be notified.")
    entry.status = "notified"
    entry.notifiedAt = now_iso()
    entry.returnByAt = iso_at_offset(repo.venue.gracePeriodMinutes)
    notification = Notification(
        id=repo.next_id("notif"),
        entryId=entry.id,
        recipientType="guest",
        channel="sms",
        templateType="table_ready",
        renderedMessage=repo.venue.messages.tableReadyMessage,
        deliveryStatus="sent",
        createdAt=now_iso(),
        sentAt=now_iso(),
    )
    repo.notifications.append(notification)
    notifier.send(notification)
    return entry, notification


def return_to_waiting(repo: InMemoryRepository, entry_id: str) -> WaitlistEntry:
    entry = find_entry_or_404(repo, entry_id)
    if entry.status != "notified":
        raise ServiceError(400, "invalid_status", "Only notified entries can return to waiting.")
    entry.status = "waiting"
    entry.notifiedAt = None
    entry.returnByAt = None
    return entry


def extend_return_by(repo: InMemoryRepository, entry_id: str, additional_minutes: int) -> WaitlistEntry:
    entry = find_entry_or_404(repo, entry_id)
    if entry.status != "notified" or entry.returnByAt is None:
        raise ServiceError(400, "invalid_status", "Only notified entries have a return-by time.")
    entry.returnByAt = (
        datetime.fromisoformat(entry.returnByAt) + timedelta(minutes=additional_minutes)
    ).isoformat()
    return entry


def update_wait_estimate(repo: InMemoryRepository, entry_id: str, estimated_wait_minutes: int) -> WaitlistEntry:
    entry = find_entry_or_404(repo, entry_id)
    if estimated_wait_minutes < 0:
        raise ServiceError(400, "invalid_estimate", "Estimated wait must be a whole number of minutes.")
    entry.estimatedWaitMinutes = estimated_wait_minutes
    return entry


def seat_entry(repo: InMemoryRepository, entry_id: str, payload: SeatEntryRequest) -> WaitlistEntry:
    entry = find_entry_or_404(repo, entry_id)
    if entry.status in {"cancelled", "no_show", "completed", "seated"}:
        raise ServiceError(400, "invalid_status", "This entry cannot be seated.")
    table = repo.find_table(payload.tableId)
    if table is None:
        raise ServiceError(404, "table_not_found", "That table does not exist.")
    if not table.active:
        raise ServiceError(400, "table_inactive", f"{table.name} is inactive.")
    if table.availabilityState == "occupied":
        raise ServiceError(400, "table_occupied", f"{table.name} is already occupied.")
    if table.availabilityState == "needs_tidying":
        raise ServiceError(
            400, "table_not_available", f"{table.name} needs tidying before it can be seated."
        )
    if table.maxCapacity < entry.partySize:
        raise ServiceError(
            400,
            "table_too_small",
            f"{table.name} seats up to {table.maxCapacity} and this party is {entry.partySize}.",
        )
    if table.minCapacity > entry.partySize:
        raise ServiceError(
            400,
            "table_capacity_mismatch",
            f"{table.name} seats a minimum of {table.minCapacity} and this party is {entry.partySize}.",
        )

    entry.status = "seated"
    entry.seatedAt = now_iso()
    entry.tableId = table.id
    entry.tableName = table.name
    entry.seatingOverrideReason = (
        payload.seatingOverrideReason.strip() if payload.seatingOverrideReason and payload.seatingOverrideReason.strip() else None
    )
    table.availabilityState = "occupied"
    table.occupyingEntryId = entry.id
    table.occupyingTicketCode = entry.ticketCode
    return entry


def mark_no_show(repo: InMemoryRepository, entry_id: str) -> WaitlistEntry:
    entry = find_entry_or_404(repo, entry_id)
    if entry.status != "notified":
        raise ServiceError(400, "invalid_status", "Only notified guests can be marked no-show.")
    entry.status = "no_show"
    return entry


def complete_entry(repo: InMemoryRepository, entry_id: str) -> WaitlistEntry:
    entry = find_entry_or_404(repo, entry_id)
    if entry.status != "seated":
        raise ServiceError(400, "invalid_status", "Only seated parties can be completed.")
    _release_table(repo, entry)
    entry.status = "completed"
    entry.completedAt = now_iso()
    return entry


def list_compatible_tables(repo: InMemoryRepository, entry_id: str) -> list[Table]:
    entry = find_entry_or_404(repo, entry_id)
    return [t for t in repo.tables if is_table_compatible(t, entry.partySize)]
